src/Core/src/Platform/iOS/KeyboardAutoManagerScroll.cs

```diff
@@ -0,0 +1,614 @@
+			return new CGRect(x, y, width, height);
+		}
+
+		return null;
+	}
+
+	static async Task SetUpTextEdit()
+	{
+		if (View is null)
+			return;
+
+		CursorRect = null;
+
+		RootController = View.FindResponder<ContainerViewController>()?.View;
+
+		// the cursor needs a small amount of time to update the position
+		await Task.Delay(5);
+
+		UITextRange? selectedTextRange;
+		CGRect? localCursor = null;
+
+		if (View is UITextView tv)
+		{
+			selectedTextRange = tv.SelectedTextRange;
+			if (selectedTextRange is UITextRange selectedRange)
+			{
+				localCursor = tv.GetCaretRectForPosition(selectedRange.Start);
+				if (localCursor is CGRect local)
+					CursorRect = tv.ConvertRectToView(local, null);
+			}
+		}
+		else if (View is UITextField tf)
+		{
+			selectedTextRange = tf.SelectedTextRange;
+			if (selectedTextRange is UITextRange selectedRange)
+			{
+				localCursor = tf.GetCaretRectForPosition(selectedRange.Start);
+				if (localCursor is CGRect local)
+					CursorRect = tf.ConvertRectToView(local, null);
+			}
+		}
+
+		TextViewTopDistance = localCursor is CGRect cGRect ? 20 + (int)cGRect.Height : 20;
+
+		await AdjustPositionDebounce();
+	}
+
+	// Used to debounce calls from different oberservers so we can be sure
+	// all the fields are updated before calling AdjustPostition()
+	internal static async Task AdjustPositionDebounce()
+	{
+		DebounceCount++;
+		var entranceCount = DebounceCount;
+
+		await Task.Delay(10);
+
+		if (entranceCount == DebounceCount)
+		{
+			AdjustPosition();
+			DebounceCount = 0;
+		}
+	}
+
+	// main method to calculate and animate the scrolling
+	internal static void AdjustPosition()
+	{
+		if (View is not UITextField field && View is not UITextView)
+			return;
+
+		if (RootController is null)
+			return;
+
+		var rootViewOrigin = new CGPoint(RootController.Frame.GetMinX(), RootController.Frame.GetMinY());
+		var window = RootController.Window;
+
+		var kbSize = KeyboardFrame.Size;
+		var intersectRect = CGRect.Intersect(KeyboardFrame, window.Frame);
+		if (intersectRect == CGRect.Empty)
+			kbSize = new CGSize(KeyboardFrame.Width, 0);
+		else
+			kbSize = intersectRect.Size;
+
+		nfloat statusBarHeight;
+		nfloat navigationBarAreaHeight;
+
+		if (RootController.GetNavigationController() is UINavigationController navigationController)
+		{
+			navigationBarAreaHeight = navigationController.NavigationBar.Frame.GetMaxY();
+		}
+		else
+		{
+			if (OperatingSystem.IsIOSVersionAtLeast(13, 0))
+				statusBarHeight = window.WindowScene?.StatusBarManager?.StatusBarFrame.Height ?? 0;
+			else
+				statusBarHeight = UIApplication.SharedApplication.StatusBarFrame.Height;
+
+			navigationBarAreaHeight = statusBarHeight;
+		}
+
+		var topLayoutGuide = Math.Max(navigationBarAreaHeight, RootController.LayoutMargins.Bottom) + 5;
+
+		var keyboardYPosition = window.Frame.Height - kbSize.Height - TextViewTopDistance;
+
+		CGRect cursorRect;
+
+		if (CursorRect is CGRect cRect)
+			cursorRect = cRect;
+		else
+			return;
```

Comment:
Inverted if looks nicer.
```suggestion
		if (CursorRect is not CGRect cRect)
			return;
		cursorRect = cRect; 
```

Reply:
Inverted does not get the type checking though which is needed for the other uses of cursorRect

Reply:
Can you elaborate, is not should be assigning cRect as a CGrect.

Reply:
I mean that CursorRect is of type CGRect? and I want to assign the cursorRect to be just the CGRect type (non nullable). So I am checking that it is of type CGRect non-nullable and then assigning it to the cursorRect variable that I use later.